### PR TITLE
[hail] Add packed integer EType

### PR DESCRIPTION
**Pull request by chrisvittal**
We use Lemire et al.'s Stream VByte algorithm, as described in 
https://arxiv.org/abs/1709.08990

```
Encoded Format:
  int (length)
  missing bits (if necessary)
  int (length of control bytes)
  int (length of data bytes)
  control bytes
  data bytes
```

This was split off from continuted development on top of #7803, ideally they should be merged together.

### CODE REVIEW

**Comment by chrisvittal** [comment]
If we want to, I can split https://github.com/hail-is/hail/pull/7821/commits/a28d336e77f9c851824ba5c0494427141b676266 and https://github.com/hail-is/hail/pull/7821/commits/d766212703e60d310dc79d139d9cb2bf3c16bfe4 out into a separate PR as those are just cleanups/fixes.

**Comment by akotlar** [comment]
This is really cool Chris

**Comment by tpoterba** [comment]
Let's chat today to make a plan for getting these reviewed. I certainly want to see size + performance benchmarks -- I'm more than a little concerned that this change by itself will be *slower* for the 1kg matrix tables -- you'll do several java object allocations per matrix entry to decode with these (for AD and PL arrays), where previously we were doing none.

**Comment by chrisvittal** [comment]
This isn't great, maybe LEB isn't _that bad_.
```
$ du -s bench-pr/profile25/*.mt bench-master/profile25/*.mt
300940  bench-pr/profile25/profile.mt
245532  bench-master/profile25/profile.mt
```
This PR:
```
2020-01-14 12:48:12,999: INFO: [1/2] Running matrix_table_decode_and_count...
2020-01-14 12:48:27,406: INFO: burn in: 14.40s
2020-01-14 12:48:39,355: INFO: run 1: 11.95s
2020-01-14 12:48:51,437: INFO: run 2: 12.08s
2020-01-14 12:49:03,650: INFO: run 3: 12.21s
2020-01-14 12:49:03,650: INFO: [2/2] Running matrix_table_decode_and_count_just_gt...
2020-01-14 12:49:07,804: INFO: burn in: 4.15s
2020-01-14 12:49:11,877: INFO: run 1: 4.07s
2020-01-14 12:49:15,825: INFO: run 2: 3.95s
2020-01-14 12:49:19,756: INFO: run 3: 3.93s
```
Master:
```
2020-01-14 12:46:17,284: INFO: [1/2] Running matrix_table_decode_and_count...
2020-01-14 12:46:23,910: INFO: burn in: 6.62s
2020-01-14 12:46:29,142: INFO: run 1: 5.23s
2020-01-14 12:46:34,365: INFO: run 2: 5.22s
2020-01-14 12:46:39,507: INFO: run 3: 5.14s
2020-01-14 12:46:39,507: INFO: [2/2] Running matrix_table_decode_and_count_just_gt...
2020-01-14 12:46:44,000: INFO: burn in: 4.49s
2020-01-14 12:46:48,074: INFO: run 1: 4.07s
2020-01-14 12:46:52,106: INFO: run 2: 4.03s
2020-01-14 12:46:56,300: INFO: run 3: 4.19s
```

**Comment by chrisvittal** [comment]
Ok, a little better now:
```
du -sm /var/tmp/hail/bench-pr/profile25/profile.mt /var/tmp/hail/bench-master/profile25/profile.mt
290	/var/tmp/hail/bench-pr/profile25/profile.mt
240	/var/tmp/hail/bench-master/profile25/profile.mt

PR:
2020-01-14 17:04:10,439: INFO: Done writing profile MT
2020-01-14 17:04:11,200: INFO: [1/2] Running matrix_table_decode_and_count...
2020-01-14 17:04:21,161: INFO: burn in: 9.96s
2020-01-14 17:04:30,477: INFO: run 1: 9.31s
2020-01-14 17:04:39,140: INFO: run 2: 8.66s
2020-01-14 17:04:48,039: INFO: run 3: 8.90s
2020-01-14 17:04:48,039: INFO: [2/2] Running matrix_table_decode_and_count_just_gt...
2020-01-14 17:04:51,293: INFO: burn in: 3.25s
2020-01-14 17:04:54,337: INFO: run 1: 3.04s
2020-01-14 17:04:57,518: INFO: run 2: 3.18s
2020-01-14 17:05:00,481: INFO: run 3: 2.96s

Master:
2020-01-14 17:05:13,828: INFO: [1/2] Running matrix_table_decode_and_count...
2020-01-14 17:05:19,819: INFO: burn in: 5.99s
2020-01-14 17:05:24,747: INFO: run 1: 4.93s
2020-01-14 17:05:29,434: INFO: run 2: 4.68s
2020-01-14 17:05:34,013: INFO: run 3: 4.58s
2020-01-14 17:05:34,013: INFO: [2/2] Running matrix_table_decode_and_count_just_gt...
2020-01-14 17:05:38,027: INFO: burn in: 4.01s
2020-01-14 17:05:41,827: INFO: run 1: 3.80s
2020-01-14 17:05:45,537: INFO: run 2: 3.71s
2020-01-14 17:05:49,339: INFO: run 3: 3.80s
```

**Comment by chrisvittal** [comment]
The one thing we probably need is the ability to choose to use this encoding as a hidden option.

File: hail/src/main/scala/is/hail/expr/types/encoded/EType.scala [hunk 1]
```diff
@@ -256,6 +256,9 @@ object EType {
       case t: PFloat64 => EFloat64(t.required)
       case t: PBoolean => EBoolean(t.required)
       case t: PBinary => EBinary(t.required)
+      // FIXME(chrisvittal): turn this on when performance is adequate
```
Comment (chrisvittal) [inline hail/src/main/scala/is/hail/expr/types/encoded/EType.scala]: We need a hidden option to turn this on for perf and correctness testing.

**Comment by tpoterba** [comment]
HailContext flag, maybe?

**Comment by chrisvittal** [comment]
Can confirm that at least locally, the environment variable caused a table to be written with packed integers, and the tests passed with the flag on (I assume).

**Comment by tpoterba** [comment]
once this is in, am I right that the plan is:
- get transposed array<struct> etype in similarly
- make transposed array<struct> ptype
- optimize these three in tandem to beat current performance / file size

**Pull request merged**